src/flp/test_types.rs

```diff
@@ -0,0 +1,112 @@
+use crate::field::FieldElement;
+use crate::flp::{FlpError, Gadget, Type};
+use std::convert::{TryFrom, TryInto};
+
+/// The average type. Each measurement is an integer in `[0,2^bits)` for some `0 < bits < 64` and the
+/// aggregate is the arithmetic average. This type is a test for demonstrating that the number of measurements
+/// may be used in the `decode_result()` function. No validity checks are done on the input data.
+///
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Average<F: FieldElement> {
+    bits: usize,
+    one: F::Integer,
+}
+
+impl<F: FieldElement> Average<F> {
+    pub fn new(bits: usize) -> Result<Self, FlpError> {
+        let bits_int = F::Integer::try_from(bits).map_err(|err| {
+            FlpError::Encode(format!(
+                "bit length ({}) cannot be represented as a field element: {:?}",
+                bits, err,
+            ))
+        })?;
+
+        if F::modulus() >> bits_int == F::Integer::from(F::zero()) {
+            return Err(FlpError::Encode(format!(
+                "bit length ({}) exceeds field modulus",
+                bits,
+            )));
+        }
+
+        let one = F::Integer::from(F::one());
+
+        Ok(Self { bits, one })
+    }
+}
+
+impl<F: FieldElement> Type for Average<F> {
+    type Measurement = F::Integer;
+    type AggregateResult = f64;
+    type Field = F;
+
+    fn encode_measurement(&self, summand: &F::Integer) -> Result<Vec<F>, FlpError> {
+        let mut encoded: Vec<F> = Vec::with_capacity(self.bits);
+        for l in 0..self.bits {
+            let l = F::Integer::try_from(l).unwrap();
+            let w = F::from((*summand >> l) & self.one);
+            encoded.push(w);
+        }
+
+        Ok(encoded)
+    }
+
+    fn decode_result(&self, data: &[F], num_measurements: usize) -> Result<f64, FlpError> {
+        // Compute the average from the aggregated sum.
+        let data = F::Integer::from(data[0]);
+        let data: u64 = data.try_into().unwrap();
+        let result = (data as f64) / (num_measurements as f64);
+        Ok(result)
+    }
+
+    fn gadget(&self) -> Vec<Box<dyn Gadget<F>>> {
+        vec![]
+    }
+
+    fn valid(
+        &self,
+        _g: &mut Vec<Box<dyn Gadget<F>>>,
+        _input: &[F],
+        _joint_rand: &[F],
+        _num_shares: usize,
+    ) -> Result<F, FlpError> {
+        // Accept everything since this is only a test type.
```

Comment:
This looks to me like a critical vulnerability waiting too happen. Imagine someone comes along and enablers `Prio3Aes128Average` without thoroughly checking the code, or uses `Average` for some other purpose (it's in the public API afterall).

In my view, this `Type` should be implemented properly, especially those in the public API. It shouldn't be too bad if we refactor the code to crib off of `Sum`.